### PR TITLE
Add Xcode build script for macOS target

### LINKED ISSUE #31274: Crashing during hot reload when --track-widget-creation is enabled
If I run a macOS Flutter application in VS Code with the new desktop `flutter run` functionality, hot reload crashes the target app; some searching based on the stack turned up https://github.com/Dart-Code/Dart-Code/issues/1133 and related bugs, which it turned out is the culprit here as well (disabling widget creation tracking fixes the problem).

@DanTup @devoncarew can you provide some more context on what the underlying fix involved? Is there per-platform support that needs to be wired up for this feature in Flutter?

(Note: I haven't tried yet on Linux and Windows; it may happen there as well.)

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Removes indirection by moving the macOS build script to flutter_tools.  This still uses the FDE asset syncing. This uses almost the same scheme/flag passing setup as the existing iOS build. 

~~Currently the track-widget-creation flag is still broken though, it seems like some part of the process is  using the regular .dill instead of the .track.dill - still investigating.~~

Fixed

See also: https://github.com/google/flutter-desktop-embedding/pull/336

## Related Issues

Fixes https://github.com/flutter/flutter/issues/31274

https://github.com/flutter/flutter/issues/30707
https://github.com/flutter/flutter/issues/30706

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behaviour with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/build_bundle.dart [hunk 1]
```diff
@@ -28,7 +28,7 @@ class BuildBundleCommand extends BuildSubCommand {
       ..addOption('depfile', defaultsTo: defaultDepfilePath)
       ..addOption('target-platform',
         defaultsTo: 'android-arm',
-        allowed: <String>['android-arm', 'android-arm64', 'android-x86', 'android-x64', 'ios'],
+        allowed: <String>['android-arm', 'android-arm64', 'android-x86', 'android-x64', 'ios', 'darwin-x64'],
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/commands/build_bundle.dart]: Shouldn't having this in the allowed list be gated behind the desktop-enabled check?

File: packages/flutter_tools/test/commands/build_macos_test.dart [hunk 2]
```diff
@@ -64,14 +65,22 @@ void main() {
...
     final BuildCommand command = BuildCommand();
     applyMocksToCommand(command);
-    fs.file('macos/build.sh').createSync(recursive: true);
+    fs.directory('macos').createSync();
     fs.file('pubspec.yaml').createSync();
     fs.file('.packages').createSync();
+    final FlutterProject flutterProject = await FlutterProject.fromDirectory(fs.currentDirectory);
+    final Directory flutterBuildDir = fs.directory(getMacOSBuildDirectory());
+
     when(mockProcessManager.start(<String>[
```
Comment (stuartmorgan) [inline packages/flutter_tools/test/commands/build_macos_test.dart]: TODO to refactor eventually to share with `ios/mac.dart`?

**Review by stuartmorgan-g — COMMENTED** [review]
I haven't had a chance to try this out locally yet (will do that tomorrow, and then re-review) but looks generally good so far.

**Review by stuartmorgan-g — APPROVED** [review]

File: packages/flutter_tools/lib/src/commands/build_bundle.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/build_bundle.dart]: Done

File: packages/flutter_tools/bin/macos_build_flutter_assets.sh [hunk 3]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
+# Copyright 2018 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/bin/macos_build_flutter_assets.sh]: Done

File: packages/flutter_tools/test/commands/build_macos_test.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/commands/build_macos_test.dart]: Done, in build_macos.dart

**Pull request merged**